// @tsed/platform-express/lib/services/PlatformExpressApplication.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlatformExpressApplication = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@tsed/common");
const Express = require("express");
const PlatformExpressRouter_1 = require("./PlatformExpressRouter");
/**
 * @platform
 * @express
 */
let PlatformExpressApplication = class PlatformExpressApplication extends PlatformExpressRouter_1.PlatformExpressRouter {
    constructor(platformHandler, configuration) {
        super(platformHandler, configuration, {
            mergeParams: true
        });
        this.rawApp = configuration.get("express.app") || Express();
        this.useContext().getApp().use(this.getRouter());
    }
    getApp() {
        return this.rawApp;
    }
    callback() {
        return this.rawApp;
    }
    useContext() {
        this.getApp().use(async (req, res, next) => {
            await common_1.createContext(this.injector, req, res);
            return next();
        });
        return this;
    }
};
PlatformExpressApplication = tslib_1.__decorate([
    tslib_1.__param(0, common_1.Inject()), tslib_1.__param(1, common_1.Configuration()),
    tslib_1.__metadata("design:paramtypes", [common_1.PlatformHandler, Object])
], PlatformExpressApplication);
exports.PlatformExpressApplication = PlatformExpressApplication;
