// @tsed/common/lib/mvc/models/ParamTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParamTypes = void 0;
var ParamTypes;
(function (ParamTypes) {
    ParamTypes["NEXT_FN"] = "NEXT_FN";
    ParamTypes["ERR"] = "ERR";
    ParamTypes["REQUEST"] = "REQUEST";
    ParamTypes["PLATFORM_REQUEST"] = "PLATFORM_REQUEST";
    ParamTypes["NODE_REQUEST"] = "NODE_REQUEST";
    ParamTypes["RESPONSE"] = "RESPONSE";
    ParamTypes["PLATFORM_RESPONSE"] = "PLATFORM_RESPONSE";
    ParamTypes["NODE_RESPONSE"] = "NODE_RESPONSE";
    ParamTypes["RESPONSE_DATA"] = "RESPONSE_DATA";
    ParamTypes["ENDPOINT_INFO"] = "ENDPOINT_INFO";
    ParamTypes["$CTX"] = "$CTX";
    ParamTypes["RES"] = "RES";
    ParamTypes["REQ"] = "REQ";
    ParamTypes["BODY"] = "BODY";
    ParamTypes["PATH"] = "PATH";
    ParamTypes["QUERY"] = "QUERY";
    ParamTypes["HEADER"] = "HEADER";
    ParamTypes["COOKIES"] = "COOKIES";
    ParamTypes["LOCALS"] = "LOCALS";
    ParamTypes["SESSION"] = "SESSION";
    ParamTypes["FILES"] = "FILES";
})(ParamTypes = exports.ParamTypes || (exports.ParamTypes = {}));
