// @tsed/passport/lib/PassportModule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PassportModule = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@tsed/common");
const Passport = require("passport");
const PassportSerializerService_1 = require("./services/PassportSerializerService");
const ProtocolsService_1 = require("./services/ProtocolsService");
/**
 * @ignore
 */
let PassportModule = class PassportModule {
    constructor(app, protocolsService, passportSerializer) {
        this.app = app;
        this.protocolsService = protocolsService;
        this.passportSerializer = passportSerializer;
    }
    $onInit() {
        Passport.serializeUser(this.passportSerializer.serialize.bind(this.passportSerializer));
        Passport.deserializeUser(this.passportSerializer.deserialize.bind(this.passportSerializer));
        this.protocolsService.getProtocols().forEach((provider) => this.protocolsService.invoke(provider));
        return undefined;
    }
    $beforeRoutesInit() {
        const { userProperty, pauseStream } = this;
        switch (this.platformName) {
            case "express":
                this.app.use(Passport.initialize({ userProperty }));
                this.app.use(Passport.session({ pauseStream }));
                return;
            default:
                console.warn(`Platform "${this.platformName}" not supported by @tsed/passport`);
        }
    }
};
tslib_1.__decorate([
    common_1.Constant("passport.userProperty"),
    tslib_1.__metadata("design:type", String)
], PassportModule.prototype, "userProperty", void 0);
tslib_1.__decorate([
    common_1.Constant("passport.pauseStream"),
    tslib_1.__metadata("design:type", Boolean)
], PassportModule.prototype, "pauseStream", void 0);
tslib_1.__decorate([
    common_1.Constant("PLATFORM_NAME"),
    tslib_1.__metadata("design:type", String)
], PassportModule.prototype, "platformName", void 0);
PassportModule = tslib_1.__decorate([
    common_1.Module({
        scope: common_1.ProviderScope.SINGLETON
    }),
    tslib_1.__metadata("design:paramtypes", [common_1.PlatformApplication,
        ProtocolsService_1.ProtocolsService,
        PassportSerializerService_1.PassportSerializerService])
], PassportModule);
exports.PassportModule = PassportModule;
