// @tsed/common/lib/mvc/decorators/method/returnType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReturnsArray = exports.Returns = exports.ReturnType = void 0;
const schema_1 = require("@tsed/schema");
/**
 * @ignore
 */
function mapStatusResponseOptions(args) {
    const configuration = {};
    args.forEach((value) => {
        configuration[typeof value] = value;
    });
    const { number: code, object: options = {}, function: type } = configuration;
    if (type) {
        options.type = type;
    }
    return {
        ...options,
        code,
        type: options.type || options.use,
        collectionType: options.collectionType || options.collection
    };
}
/**
 * Define the returned type for the serialization.
 *
 * ::: warning
 * This decorator will be removed in v6 in favor of @@Returns@@ from @tsed/schema.
 * For v5 user, use @@Returns@@ decorator from @tsed/common then in v6 switch to @tsed/schema.
 * :::
 *
 * ```typescript
 * @Controller('/')
 * export class Ctrl {
 *    @Get('/')
 *    @ReturnType({code: 200, type: User, collectionType: Map})
 *    get(): Promise<Map<User>> { }
 * }
 *
 * ```
 *
 * @returns {Function}
 * @param response
 * @decorator
 * @operation
 * @response
 * @deprecated Since v6. Use @Returns decorator from @tsed/schema
 * @ignore
 */
function ReturnType(response = {}) {
    const { code = "default", collectionType, type, headers, description, examples, schema } = response;
    let decorator = schema_1.Returns(code);
    if (collectionType || type) {
        decorator.Type(collectionType || type);
    }
    if (collectionType) {
        decorator = decorator.Of(type);
    }
    if (headers) {
        decorator = decorator.Headers(headers);
    }
    if (description) {
        decorator = decorator.Description(description);
    }
    if (examples) {
        decorator = decorator.Examples(examples);
    }
    if (schema) {
        decorator = decorator.Schema(schema);
    }
    return decorator;
}
exports.ReturnType = ReturnType;
/**
 * @deprecated Since v6. Use @Returns decorator from @tsed/schema
 * @ignore
 */
function Returns(...args) {
    return ReturnType(mapStatusResponseOptions(args));
}
exports.Returns = Returns;
/**
 * @deprecated Since v6. Use @Returns decorator from @tsed/schema
 */
function ReturnsArray(...args) {
    return ReturnType({ ...mapStatusResponseOptions(args), collectionType: Array });
}
exports.ReturnsArray = ReturnsArray;
