// @tsed/core/lib/domain/Entity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Entity = void 0;
const utils_1 = require("../utils");
class Entity {
    constructor({ target, propertyKey, descriptor, index }) {
        this._target = target;
        this.propertyKey = propertyKey;
        this.propertyName = String(propertyKey);
        this.descriptor = descriptor;
        this.index = index;
        this.decoratorType = utils_1.decoratorTypeOf([target, propertyKey, descriptor || index]);
        this.token = target && utils_1.classOf(target);
    }
    /**
     * Reference to the class
     */
    get target() {
        return this._target;
    }
    /**
     * Return the class name of the entity.
     * @returns {string}
     */
    get targetName() {
        return utils_1.nameOf(this.token);
    }
    /**
     * Return the collection name
     * @returns {string}
     * @todo should not be use in final API
     */
    get collectionName() {
        return this.collectionType ? utils_1.nameOf(this.collectionType) : "";
    }
    /**
     *
     * @returns {Type<any>}
     */
    get type() {
        return this._type;
    }
    /**
     * Get original type without transformation
     * @param value
     */
    set type(value) {
        this._type = value;
    }
    /**
     * Return target type name
     * @returns {string}
     * @todo should not be use in final API
     */
    get typeName() {
        return utils_1.nameOf(this._type);
    }
    /**
     * Return the itemSchema target type. if the type is a function used for recursive model, the function will be called to
     * get the right type.
     */
    get computedType() {
        return this._type;
    }
    /**
     *
     * @returns {boolean}
     */
    get isCollection() {
        return !!this.collectionType;
    }
    /**
     *
     * @returns {boolean}
     */
    get isArray() {
        return utils_1.isArrayOrArrayClass(this.collectionType);
    }
    /**
     *
     * @returns {boolean}
     */
    get isPrimitive() {
        return utils_1.isPrimitiveOrPrimitiveClass(this._type);
    }
    /**
     *
     * @returns {boolean}
     */
    get isDate() {
        return utils_1.isDate(this.computedType);
    }
    /**
     *
     * @returns {boolean}
     */
    get isObject() {
        return utils_1.isObject(this.computedType);
    }
    /**
     *
     */
    get isClass() {
        return utils_1.isClass(this.computedType);
    }
}
exports.Entity = Entity;
