// @tsed/mongoose/lib/services/MongooseService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MongooseService = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@tsed/common");
const Mongoose = require("mongoose");
let MongooseService = class MongooseService {
    constructor() {
        this.connections = new Map();
        this.defaultConnection = "default";
    }
    /**
     *
     * @returns {Promise<"mongoose".Connection>}
     */
    async connect(id, url, connectionOptions, isDefault = false) {
        if (this.has(id)) {
            return await this.get(id);
        }
        common_1.$log.info(`Connect to mongo database: ${id}`);
        common_1.$log.debug(`Url: ${url}`);
        common_1.$log.debug(`options: ${JSON.stringify(connectionOptions)}`);
        try {
            const connection = await Mongoose.createConnection(url, connectionOptions);
            this.connections.set(id, connection);
            if (id === "default" || isDefault) {
                this.defaultConnection = id;
            }
            return connection;
        }
        catch (er) {
            /* istanbul ignore next */
            common_1.$log.error(er);
            /* istanbul ignore next */
            process.exit();
        }
    }
    /**
     *
     * @returns {"mongoose".Connection}
     */
    get(id) {
        return this.connections.get(id || this.defaultConnection);
    }
    /**
     *
     * @param {string} id
     * @returns {boolean}
     */
    has(id) {
        return this.connections.has(id || this.defaultConnection);
    }
    async closeConnections() {
        for (const [id, connection] of this.connections.entries()) {
            await connection.close();
            this.connections.delete(id);
        }
    }
};
MongooseService = tslib_1.__decorate([
    common_1.Service()
], MongooseService);
exports.MongooseService = MongooseService;
